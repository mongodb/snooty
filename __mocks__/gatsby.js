const gatsby = jest.requireActual('gatsby');

module.exports = {
  ...gatsby,
  graphql: jest.fn(),
  StaticQuery: jest.fn(),
  withPrefix: jest.fn().mockImplementation(str => str),
  useStaticQuery: jest.fn(),
};
