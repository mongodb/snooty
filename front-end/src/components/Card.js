import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Card extends Component {
  getTitle(val) {
    const { refDocMapping } = this.props;
    if (refDocMapping[val] && refDocMapping[val].ast) {
      return refDocMapping[val].ast.children[0].children[0].children[0].value;
    }
    return 'Title not found';
  }

  cardContent() {
    const { card, time } = this.props;
    const innerContent = (
      <React.Fragment>
        <div className="guide__title">
          {card.name === 'card' ? this.getTitle(card.argument[0].value) : card.argument[0].value}
        </div>
        {card.name === 'multi-card' ? (
          <ul className="guide__body">
            {card.children[0].children.map((listItem, index) => (
              <li className="guide__entry" key={index}>
                <a href={listItem.children[0].children[0].value}>
                  {this.getTitle(listItem.children[0].children[0].value)}
                </a>
              </li>
            ))}
          </ul>
        ) : (
          <div className="guide__body" />
        )}
        <ul className="guide__pills" />
        {card.name === 'card' && <div className="guide__time">{time} min</div>}
      </React.Fragment>
    );
    if (card.name === 'multi-card') {
      return <div className="guide guide--jumbo guide--expanded">{innerContent}</div>;
    }
    return (
      <a href={card.argument[0].value} className="guide guide--regular">
        {innerContent}
      </a>
    );
  }

  render() {
    return this.cardContent();
  }
}

Card.propTypes = {
  card: PropTypes.shape({
    argument: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string,
      })
    ),
    children: PropTypes.arrayOf(
      PropTypes.shape({
        children: PropTypes.array,
      })
    ),
    name: PropTypes.string.isRequired,
  }).isRequired,
  refDocMapping: PropTypes.shape({
    index: PropTypes.shape({
      ast: PropTypes.object,
    }).isRequired,
  }).isRequired,
  time: PropTypes.string,
};

Card.defaultProps = {
  time: undefined,
};
