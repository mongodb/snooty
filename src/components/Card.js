import React from 'react';
import PropTypes from 'prop-types';
import { findKeyValuePair } from '../util';
import { stringifyTab } from '../constants';
import { setLocalValue } from '../localStorage';

const CardPills = ({ pillsNode, pillsetName }) => {
  const getPillTitle = node => node.children[0].children[0].value;

  // Get the name of each pill. If the total character length of pills > 50, truncate them and add a "See All" pill.
  const mapPills = node => {
    let totalLength = 0;
    let isTruncated = false;
    const pills = [];
    node.children[0].children.forEach(pillObj => {
      const pill = getPillTitle(pillObj);
      if (totalLength > 50) {
        isTruncated = true;
      } else {
        pills.push(pill);
      }
      totalLength += stringifyTab(pill).length + 4;
    });
    return [pills, isTruncated];
  };

  const [pills, isTruncated] = mapPills(pillsNode);

  return (
    <ul className="guide__pills">
      {pills.map((pill, index) => (
        <li className="guide__pill" key={index}>
          <span
            onClick={() => {
              setLocalValue(pillsetName, pill);
            }}
            role="button"
            tabIndex={index}
          >
            {stringifyTab(pill)}
          </span>
        </li>
      ))}
      {isTruncated && <li className="guide__pill guide__pill--seeall">See All</li>}
    </ul>
  );
};

CardPills.propTypes = {
  pillsetName: PropTypes.string.isRequired,
  pillsNode: PropTypes.shape({
    name: PropTypes.oneOf(['languages']).isRequired,
    children: PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.oneOf(['list']).isRequired,
        children: PropTypes.arrayOf(
          PropTypes.shape({
            type: PropTypes.oneOf(['listItem']).isRequired,
            children: PropTypes.arrayOf(
              PropTypes.shape({
                type: PropTypes.oneOf(['paragraph']).isRequired,
                children: PropTypes.arrayOf(
                  PropTypes.shape({
                    type: PropTypes.oneOf(['text']).isRequired,
                    value: PropTypes.string.isRequired,
                  })
                ).isRequired,
              })
            ).isRequired,
          })
        ).isRequired,
      })
    ).isRequired,
  }).isRequired,
};

const Card = ({ card, refDocMapping, time }) => {
  const getCardTitle = cardSlug => {
    if (refDocMapping[cardSlug].ast) {
      return refDocMapping[cardSlug].ast.children[0].children[0].children[0].value;
    }
    return 'Title not found';
  };

  const getPills = cardSlug => {
    return findKeyValuePair(refDocMapping[cardSlug].ast.children, 'name', 'languages');
  };

  const cardContent = () => {
    const cardSlug = card.argument[0].value;
    const languagesNode = card.name !== 'multi-card' ? getPills(cardSlug) : undefined;
    const innerContent = (
      <React.Fragment>
        <div className="guide__title">{card.name === 'card' ? getCardTitle(cardSlug) : cardSlug}</div>
        {card.name === 'multi-card' ? (
          <ul className="guide__body">
            {card.children[0].children.map((listItem, index) => (
              <li className="guide__entry" key={index}>
                <a href={listItem.children[0].children[0].value}>
                  {getCardTitle(listItem.children[0].children[0].value)}
                </a>
              </li>
            ))}
          </ul>
        ) : (
          <div className="guide__body" />
        )}
        {card.name !== 'multi-card' && languagesNode && <CardPills pillsNode={languagesNode} pillsetName="drivers" />}
        {card.name === 'card' && <div className="guide__time">{time}min</div>}
      </React.Fragment>
    );
    if (card.name === 'multi-card') {
      return <div className="guide guide--jumbo guide--expanded">{innerContent}</div>;
    }
    return (
      <a href={card.argument[0].value} className="guide guide--regular">
        {innerContent}
      </a>
    );
  };

  return cardContent();
};

Card.propTypes = {
  card: PropTypes.shape({
    argument: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string,
      })
    ),
    children: PropTypes.arrayOf(
      PropTypes.shape({
        children: PropTypes.array,
      })
    ),
    name: PropTypes.string.isRequired,
  }).isRequired,
  refDocMapping: PropTypes.oneOfType([
    PropTypes.shape({
      index: PropTypes.shape({
        ast: PropTypes.object,
      }).isRequired,
    }),
    PropTypes.shape({
      [PropTypes.string]: PropTypes.shape({
        ast: PropTypes.object.isRequired,
      }).isRequired,
    }).isRequired,
  ]),
  time: PropTypes.string,
};

Card.defaultProps = {
  refDoc: undefined,
  time: undefined,
};

export default Card;

/* 
    [PropTypes.string]: PropTypes.shape({
      ast: PropTypes.shape({
        children: PropTypes.arrayOf(
          PropTypes.shape({
            children: PropTypes.arrayOf(
              PropTypes.shape({
                children: PropTypes.arrayOf(
                  PropTypes.shape({
                    value: PropTypes.string.isRequired,
                  })
                ).isRequired,
              })
            ).isRequired,
          })
        ).isRequired,
      }).isRequired,
    }).isRequired, */
