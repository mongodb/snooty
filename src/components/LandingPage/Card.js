import React from 'react';
import PropTypes from 'prop-types';
import CardPills from './CardPills';
import Link from '../Link';
import { getNestedValue } from '../../utils/get-nested-value';

const DEFAULT_COMPLETION_TIME = 15;

const Card = ({ card, guidesMetadata }) => {
  const getCardTitle = (cardSlug) => getNestedValue([cardSlug, 'title'], guidesMetadata);
  const getCompletionTime = (cardSlug) =>
    getNestedValue([cardSlug, 'completionTime'], guidesMetadata) || DEFAULT_COMPLETION_TIME;
  const getPills = (cardSlug) => getNestedValue([cardSlug, 'languages'], guidesMetadata);

  const cardContent = () => {
    const cardSlug = getNestedValue(['argument', 0, 'value'], card);
    const languagesNode = card.name !== 'multi-card' ? getPills(cardSlug) : undefined;
    const multiCardEntries = getNestedValue(['children', 0, 'children'], card);
    const innerContent = (
      <React.Fragment>
        <div className="guide__title">{card.name === 'card' ? getCardTitle(cardSlug) : cardSlug}</div>
        {card.name === 'multi-card' ? (
          <ul className="guide__body">
            {multiCardEntries.length > 0 &&
              multiCardEntries.map((listItem, index) => (
                <li className="guide__entry" key={index}>
                  <Link to={getNestedValue(['children', 0, 'children', 0, 'value'], listItem)}>
                    {getCardTitle(getNestedValue(['children', 0, 'children', 0, 'value'], listItem))}
                  </Link>
                </li>
              ))}
          </ul>
        ) : (
          <div className="guide__body" />
        )}
        {card.name !== 'multi-card' && languagesNode && <CardPills pillsNode={languagesNode} pillsetName="drivers" />}
        {card.name === 'card' && <div className="guide__time">{getCompletionTime(cardSlug)}min</div>}
      </React.Fragment>
    );
    if (card.name === 'multi-card') {
      return <div className="guide guide--jumbo guide--expanded">{innerContent}</div>;
    }
    return (
      <Link to={getNestedValue(['argument', 0, 'value'], card)} className="guide guide--regular">
        {innerContent}
      </Link>
    );
  };

  return cardContent();
};

Card.propTypes = {
  card: PropTypes.shape({
    argument: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string,
      })
    ),
    children: PropTypes.arrayOf(
      PropTypes.shape({
        children: PropTypes.array,
      })
    ),
    name: PropTypes.string.isRequired,
  }).isRequired,
};

Card.defaultProps = {
  refDoc: undefined,
  time: undefined,
};

export default Card;
