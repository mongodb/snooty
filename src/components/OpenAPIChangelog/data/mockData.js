export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
