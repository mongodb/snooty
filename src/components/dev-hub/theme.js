/**
 * @type {Object.<string, string>}
 * @property {string} fontsize returns px value
 */
const fontSize = {
  xsmall: '10px',
  tiny: '12px',
  small: '14px',
  default: '16px',
  h1: '48px',
  h2: '30px',
  h3: '24px',
  h4: '20px',
  tabNav: '10px',
};
/**
 * @type {Object}
 * @property {string} size returns px value
 * @property {function(string): number} stripUnit removes px unit
 */
const size = {
  tiny: '5px',
  small: '10px',
  default: '16px',
  medium: '24px',
  large: '32px',
  xlarge: '64px',
  xxlarge: '128px',
  maxWidth: '1400px',
};
const colorMap = {
  black: '#1B272D',
  darkGreen: '#13AA52',
  devBlack: '#1a282e',
  devWhite: '#e3ecea',
  greyDarkOne: '#5B6D75',
  greyDarkTwo: '#3A5058',
  greyDarkThree: '#2A3D46',
  greyLightOne: '#E3ECEA',
  greyLightTwo: '#B4C4C1',
  greyLightThree: '#85979A',
  lightGreen: '#0AD05B',
  orange: '#ED7E22',
  magenta: '#CD509B',
  salmon: '#E55F55',
  teal: '#2F9FC5',
  violet: '#6E60F9',
  white: '#FFFFFF',
  yellow: '#FDDC49',
};

/**
 * store common responsive sizes
 * @type {Object.<string, string>}
 */
const screenSize = {
  upToXSmall: 'only screen and (max-width: 374px)',
  xSmallAndUp: 'not all and (max-width: 374px)',
  upToSmall: 'only screen and (max-width: 420px)',
  smallAndUp: 'not all and (max-width: 420px)',
  upToMedium: 'only screen and (max-width: 767px)',
  mediumAndUp: 'not all and (max-width: 767px)',
  upToLarge: 'only screen and (max-width: 1023px)',
  largeAndUp: 'not all and (max-width: 1023px)',
};

export { colorMap, fontSize, screenSize, size };
