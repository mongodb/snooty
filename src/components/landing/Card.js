import React from 'react';
import PropTypes from 'prop-types';
import { withPrefix } from 'gatsby';
import styled from '@emotion/styled';
import LeafyGreenCard from '@leafygreen-ui/card';
import ComponentFactory from '../ComponentFactory';
import ConditionalWrapper from '../ConditionalWrapper';
import Link from '../Link';
import Tag from '../Tag';

const StyledCard = styled(LeafyGreenCard)`
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: ${({ theme }) => `${theme.size.large}`};

  p:last-of-type {
    margin-bottom: 0;
  }
`;

const CardIcon = styled('img')`
  width: ${({ theme }) => `${theme.size.medium}`};
`;

const H4 = styled('h4')`
  letter-spacing: 0.5px;
  margin: ${({ compact, theme }) =>
    compact ? `0 0 ${theme.size.small}` : `${theme.size.medium} 0 ${theme.size.small} 0`};
`;

const CTA = styled('p')`
  font-weight: bold;
  margin-top: auto;
`;

const FlexTag = styled(Tag)`
  margin-right: auto;
`;

const CompactCard = styled(StyledCard)`
  align-items: flex-start;
  display: flex;
  flex-direction: row;
`;

const CompactWrapper = styled('div')`
  display: flex;
  flex-direction: column;
  height: 100%;
  margin-left: ${({ theme }) => theme.size.default};
`;

const Card = ({
  isCompact,
  isExtraCompact,
  nodeData: {
    children,
    options: { cta, headline, icon, 'icon-alt': iconAlt, tag, url },
  },
}) => {
  const Card = isCompact || isExtraCompact ? CompactCard : StyledCard;
  return (
    <Card
      onClick={() => {
        window.location.href = url;
      }}
    >
      {icon && <CardIcon src={withPrefix(icon)} alt={iconAlt} />}
      <ConditionalWrapper
        condition={isCompact || isExtraCompact}
        wrapper={(children) => <CompactWrapper>{children}</CompactWrapper>}
      >
        {tag && <FlexTag text={tag} />}
        <H4 compact={isCompact || isExtraCompact}>{headline}</H4>
        {children.map((child, i) => (
          <ComponentFactory nodeData={child} key={i} />
        ))}
        {cta && (
          <CTA>
            <Link to={url}>{cta}</Link>
          </CTA>
        )}
      </ConditionalWrapper>
    </Card>
  );
};

Card.propTypes = {
  nodeData: PropTypes.shape({
    children: PropTypes.arrayOf(PropTypes.object),
    options: PropTypes.shape({
      cta: PropTypes.string,
      headline: PropTypes.string,
      icon: PropTypes.string,
      'icon-alt': PropTypes.string,
      tag: PropTypes.string,
      url: PropTypes.string,
    }),
  }).isRequired,
};

export default Card;
