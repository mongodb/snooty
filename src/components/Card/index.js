import React from 'react';
import PropTypes from 'prop-types';
import { withPrefix, navigate } from 'gatsby';
import styled from '@emotion/styled';
import LeafyGreenCard from '@leafygreen-ui/card';
import { palette } from '@leafygreen-ui/palette';
import { Body } from '@leafygreen-ui/typography';
import { theme } from '../../theme/docsTheme';
import ComponentFactory from '../ComponentFactory';
import ConditionalWrapper from '../ConditionalWrapper';
import Link from '../Link';
import Tag from '../Tag';
import { isRelativeUrl } from '../../utils/is-relative-url';

const StyledCard = styled(LeafyGreenCard)`
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: ${theme.size.large};

  p:last-of-type {
    margin-bottom: 0;
  }
`;

const CardIcon = styled('img')`
  width: ${theme.size.medium};
`;

const H4 = styled(Body)`
  letter-spacing: 0.5px;
  margin: ${({ compact, theme }) =>
    compact ? `0 0 ${theme.size.small}` : `${theme.size.medium} 0 ${theme.size.small} 0`};
`;

const CTA = styled('p')`
  margin-top: auto;
`;

const FlexTag = styled(Tag)`
  margin-right: auto;
`;

const CompactCard = styled(StyledCard)`
  align-items: flex-start;
  flex-direction: row;
  padding: ${theme.size.large} ${theme.size.medium};
`;

const CompactIcon = styled('img')`
  margin: auto;
  width: ${theme.size.medium};
  @media ${theme.screenSize.upToSmall} {
    width: 20px;
  }
`;

const CompactIconCircle = styled('div')`
  background: ${palette.green.light3};
  border-radius: 50%;
  display: flex;
  flex-shrink: 0 !important;
  height: 48px;
  width: 48px;
  @media ${theme.screenSize.upToSmall} {
    height: 40px;
    width: 40px;
  }
`;

const CompactTextWrapper = styled('div')`
  display: flex;
  flex-direction: column;
  height: 100%;
  margin-left: ${theme.size.medium};
  @media ${theme.screenSize.upToSmall} {
    margin-left: ${theme.size.default};
  }
`;

const onCardClick = (url) => {
  isRelativeUrl(url) ? navigate(url) : (window.location.href = url);
};

const Card = ({
  isCompact,
  isExtraCompact,
  nodeData: {
    children,
    options: { cta, headline, icon, 'icon-alt': iconAlt, tag, url },
  },
}) => {
  const Card = isCompact || isExtraCompact ? CompactCard : StyledCard;
  const Icon = isCompact ? CompactIcon : CardIcon;
  return (
    <Card onClick={url ? () => onCardClick(url) : undefined}>
      {icon && (
        <ConditionalWrapper
          condition={isCompact}
          wrapper={(children) => <CompactIconCircle>{children}</CompactIconCircle>}
        >
          <Icon src={withPrefix(icon)} alt={iconAlt} />
        </ConditionalWrapper>
      )}
      <ConditionalWrapper
        condition={isCompact || isExtraCompact}
        wrapper={(children) => <CompactTextWrapper>{children}</CompactTextWrapper>}
      >
        {tag && <FlexTag>{tag}</FlexTag>}
        <H4 compact={isCompact || isExtraCompact} weight="medium">
          {headline}
        </H4>
        {children.map((child, i) => (
          // The cardRef prop's purpose to distinguish wich RefRoles are coming from the Card component (a workaround while we figure out card-ref support in the parser/)
          <ComponentFactory nodeData={child} key={i} cardRef={true} />
        ))}
        {cta && (
          <CTA>
            <Link to={url}>{cta}</Link>
          </CTA>
        )}
      </ConditionalWrapper>
    </Card>
  );
};

Card.propTypes = {
  nodeData: PropTypes.shape({
    children: PropTypes.arrayOf(PropTypes.object),
    options: PropTypes.shape({
      cta: PropTypes.string,
      headline: PropTypes.string,
      icon: PropTypes.string,
      'icon-alt': PropTypes.string,
      tag: PropTypes.string,
      url: PropTypes.string,
    }),
  }).isRequired,
};

export default Card;
