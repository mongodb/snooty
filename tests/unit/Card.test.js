import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import Card from '../../src/components/Card';
import { theme } from '../../src/theme/docsTheme';
import { navigate } from 'gatsby';

// data for this component
import mockData from './data/Card.test.json';

it('renders correctly', () => {
  const tree = render(
    <ThemeProvider theme={theme}>
      <Card nodeData={mockData} />
    </ThemeProvider>
  );
  expect(tree.asFragment()).toMatchSnapshot();
});

describe('relative urls passed to Card', () => {
  test.each([
    ['/foo', 1],
    ['/foo/', 2],
    ['foo', 3],
  ])('should use Gatsby navigate function to render a link to %s', (url, expected) => {
    mockData['options']['url'] = url;
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Card nodeData={mockData} isCompact={true} />
      </ThemeProvider>
    );
    getByText('Test card headline').click();
    expect(navigate).toHaveBeenCalledTimes(expected);
  });
});
