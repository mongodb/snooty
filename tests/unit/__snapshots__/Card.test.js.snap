// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<StyledCard
  onClick={[Function]}
>
  <ConditionalWrapper
    wrapper={[Function]}
  >
    <CardIcon
      alt="MongoDB University icon"
      src="/images/icons/university.svg"
    />
  </ConditionalWrapper>
  <ConditionalWrapper
    wrapper={[Function]}
  >
    <FlexTag
      text="Test card tag"
    />
    <H4>
      Test card headline
    </H4>
    <ComponentFactory
      key="0"
      nodeData={
        Object {
          "children": Array [
            Object {
              "position": Object {
                "start": Object {
                  "line": 136,
                },
              },
              "type": "text",
              "value": "Test card paragraph",
            },
          ],
          "position": Object {
            "start": Object {
              "line": 136,
            },
          },
          "type": "paragraph",
        }
      }
    />
    <CTA>
      <Link
        to="https://university.mongodb.com/courses/M001/about"
      >
        Test card CTA
      </Link>
    </CTA>
  </ConditionalWrapper>
</StyledCard>
`;
